### PR TITLE
Add query-shape test primitives to protean.integrations.pytest

### LINKED ISSUE #953: 3.6.4: Query-shape test primitives
## Summary

Add pytest helpers that assert query shape and round-trip count: `assert_query_count(n)`, `assert_no_subquery_wrap()`, `assert_no_overfetch(expected_returned)`. These let tests catch query-cost regressions at PR time without running benchmarks.

## Problem

Performance pathologies in the query path don't surface in tests today. The outbox audit (#941) found a 3× over-fetch, a per-status loop with subquery wraps, and 1+N round trips for find-and-claim — none caught by the existing test suite. Reasons:

- Memory adapter doesn't model query cost; tests pass regardless of how many "queries" would have been issued.
- Tests use small datasets where over-fetch and round-trip count are invisible.
- No assertion primitives let test authors say "this operation should issue exactly N queries" or "this query shouldn't wrap itself in a subquery."

The result: code that performs catastrophically in production can pass every existing test.

### Available hooks

SQLAlchemy gives test authors the right primitives:

- `Engine.before_cursor_execute` event fires before every query with the SQL string.
- `Engine.after_cursor_execute` fires after with timing.
- Both can be subscribed to capture query count, query text, and execution time.

A pytest fixture wrapping these events into a counter/recorder is straightforward. The challenge is making it ergonomic — tests should read declaratively, not require event-handler setup.

### Three primitives to ship

**`assert_query_count(n)`** — context manager. Asserts the number of queries issued during the block equals `n`.

```python
def test_outbox_poll_issues_one_query(outbox_repo):
    with assert_query_count(1):
        outbox_repo.find_unprocessed(limit=10)
```

The test fails if the poll issues 2+ queries. Catches the count_by_status loop, 1+N find-and-claim, etc.

**`assert_no_subquery_wrap()`** — context manager. Inspects emitted SQL inside the block; fails if any query matches the `SELECT count(*) FROM (SELECT … FROM …) AS …` shape.

```python
def test_outbox_count_uses_flat_count(outbox_repo):
    with assert_no_subquery_wrap():
        outbox_repo.count_by_status()
```

Catches the `.limit(1).all().total` regression specifically.

**`assert_no_overfetch(expected_returned)`** — context manager. Inspects the SQL `LIMIT` clause and the actual returned row count; fails if `LIMIT > expected_returned * 1.5` (configurable ratio).

```python
def test_outbox_find_unprocessed_no_overfetch(outbox_repo, outbox_data):
    with assert_no_overfetch(expected_returned=10):
        outbox_repo.find_unprocessed(limit=10)
```

Catches the `min(limit * 3, 1000)` over-fetch pattern.

### Adapter coverage

The primitives work on the SQLAlchemy adapter (which is where production query cost matters). For memory adapter tests they're no-ops (nothing to count); for ES tests they hook the ES client's request layer. Tests using these primitives should be marked `@pytest.mark.database` or equivalent, since they only meaningfully assert against a real DB.

### Module location

`src/protean/integrations/pytest/query_assertions.py` (alongside the existing pytest integration).

### Implementation sketch

```python
# src/protean/integrations/pytest/query_assertions.py

from contextlib import contextmanager
import re
from sqlalchemy import event

@contextmanager
def assert_query_count(expected: int, engine=None):
    """Assert exactly `expected` queries are issued in the block."""
    queries = []

    def _record(conn, cursor, statement, *args, **kwargs):
        queries.append(statement)

    engine = engine or _current_engine()
    event.listen(engine, "before_cursor_execute", _record)
    try:
        yield queries
    finally:
        event.remove(engine, "before_cursor_execute", _record)
        actual = len(queries)
        if actual != expected:
            raise AssertionError(
                f"Expected {expected} queries, got {actual}.\n"
                + "\n".join(f"  {i+1}. {q}" for i, q in enumerate(queries))
            )

SUBQUERY_WRAP_PATTERN = re.compile(
    r"SELECT\s+count\(\*\)\s+FROM\s+\(SELECT\s+.+\)\s+AS\s+\w+",
    re.IGNORECASE | re.DOTALL,
)

@contextmanager
def assert_no_subquery_wrap(engine=None):
    """Fail if any query matches the SELECT count(*) FROM (...) shape."""
    offenders = []

    def _check(conn, cursor, statement, *args, **kwargs):
        if SUBQUERY_WRAP_PATTERN.search(statement):
            offenders.append(statement)

    engine = engine or _current_engine()
    event.listen(engine, "before_cursor_execute", _check)
    try:
        yield
    finally:
        event.remove(engine, "before_cursor_execute", _check)
        if offenders:
            raise AssertionError(
                f"{len(offenders)} subquery-wrapped count(s) detected:\n"
                + "\n".join(f"  - {q}" for q in offenders)
            )

@contextmanager
def assert_no_overfetch(expected_returned: int, ratio: float = 1.5, engine=None):
    """Fail if any LIMIT in the block exceeds expected_returned * ratio."""
    offenders = []
    limit_pattern = re.compile(r"\bLIMIT\s+(\d+)", re.IGNORECASE)

    def _check(conn, cursor, statement, *args, **kwargs):
        match = limit_pattern.search(statement)
        if match:
            limit = int(match.group(1))
            if limit > expected_returned * ratio:
                offenders.append((statement, limit))

    engine = engine or _current_engine()
    event.listen(engine, "before_cursor_execute", _check)
    try:
        yield
    finally:
        event.remove(engine, "before_cursor_execute", _check)
        if offenders:
            raise AssertionError(
                f"Over-fetch detected (expected ≤ {int(expected_returned * ratio)}):\n"
                + "\n".join(f"  - LIMIT {l}: {q}" for q, l in offenders)
            )
```

### Engine resolution

`_current_engine()` resolves the SQLAlchemy engine from the test's domain. For tests using the autouse `test_domain` fixture, this finds the active SQLAlchemy provider's engine; for tests with their own domain, the engine is passed explicitly.

### Outbox tests use the primitives

Once these land, the outbox test suite gains regression guards:

```python
def test_outbox_poll_no_overfetch(outbox_repo, outbox_data_500_pending):
    with assert_no_overfetch(expected_returned=10):
        outbox_repo.find_unprocessed(limit=10)

def test_outbox_count_by_status_no_subquery_wrap(outbox_repo):
    with assert_no_subquery_wrap():
        outbox_repo.count_by_status()

def test_outbox_claim_atomic_single_query(outbox_repo, outbox_data):
    with assert_query_count(1):
        outbox_repo.claim_batch(worker_id="w1", limit=10)
```

### Documentation

`docs/guides/testing-query-shape.md` — when and how to use the primitives, with examples and integration notes (`@pytest.mark.database` requirement, engine-resolution rules, false-positive guidance).

## Alternatives considered

- **SQLAlchemy `comments` extension** to inject test annotations into queries. Considered for a richer "this query was issued from test X" debugging story. Out of scope for this sub-issue; could be a follow-up.
- **Statistical regression detection** (timing-based, not shape-based). That's the microbenchmark harness's job (3.6.5). Shape primitives are deterministic and catch a different class of regression.
- **Fold into the conformance suite (3.6.1).** Considered. Rejected — conformance is about adapter contracts; query shape is about consumer correctness. Different concerns, different test placements.

## Backward compatibility

Pure addition. No existing test changes unless authors opt in.

## Test plan

- `tests/integrations/pytest/test_query_assertions.py` — the assertions themselves: a positive test (passes on correct shape) and a negative test (fails with a useful message on wrong shape) for each primitive.
- Apply primitives to the outbox test suite (in epic #941) as integration validation.

## Acceptance criteria

- [ ] `assert_query_count`, `assert_no_subquery_wrap`, `assert_no_overfetch` available from `protean.integrations.pytest`.
- [ ] Primitives work against the SQLAlchemy adapter; behave as no-ops on memory.
- [ ] Engine resolution works for tests using the autouse `test_domain` fixture and for tests with their own domain.
- [ ] `docs/guides/testing-query-shape.md` published.
- [ ] Outbox test suite uses the primitives to lock in the gains from epic #941.

### PR DESCRIPTION
**Pull request by subhashb**
## Summary

Adds three pytest context managers (issue #953) that assert query *shape* and
*round-trip count*, so query-cost regressions fail at PR time instead of in
production. Exported from `protean.integrations.pytest`:

```python
from protean.integrations.pytest import (
    assert_query_count,
    assert_no_subquery_wrap,
    assert_no_overfetch,
)
```

- **`assert_query_count(n)`** — exactly `n` data round trips (`SELECT`/`INSERT`/
  `UPDATE`/`DELETE`/`WITH`); connection setup (`PRAGMA`/`SET`) and transaction
  control are not counted, so per-connection pragmas don't inflate the number.
- **`assert_no_subquery_wrap()`** — fails on the `SELECT count(*) FROM (SELECT … )`
  shape a naive `.limit(1).all().total` emits.
- **`assert_no_overfetch(expected_returned, ratio=1.5)`** — fails when any `LIMIT`
  exceeds the rows the caller needs (checks every `LIMIT` in the statement).

They hook SQLAlchemy's `before_cursor_execute`, resolve the engine from the
active domain (or an explicit `engine=`), and are **no-ops** when no SQLAlchemy
engine is available (e.g. the in-memory adapter). `sqlalchemy` is an optional
dependency, so it is imported lazily — the auto-registered pytest plugin package
imports cleanly without it.

The outbox poll-path perf tests are **migrated onto the primitives**, replacing
their hand-rolled `capture_statements`/`sa_engine` helpers. Functional
assertions run on every backend; statement-shape assertions are gated for a real
SQL backend.

## Test plan

- [x] Primitive tests (positive + negative for each, plus no-op/no-domain and
  multi-LIMIT edge cases): green on memory / sqlite / postgresql
- [x] Outbox perf tests pass on memory / sqlite / postgresql
- [x] Core tests pass (9467)
- [x] 100% patch coverage
- [x] `import protean.integrations.pytest` does not pull in sqlalchemy

## Docs

- Guide: `docs/guides/testing/query-shape-tests.md` (+ nav, contents, how-do-i)
- Reference: query-shape assertions section in the pytest-plugin page

Closes #953

### CODE REVIEW

File: src/protean/integrations/pytest/query_assertions.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+# the engine event API is imported lazily, only on the real-engine path.
+
+# ``SELECT count(*) FROM (SELECT ... ) AS anon_1`` — a count wrapped around a
+# full subquery, the pathology the outbox audit found behind ``.limit(1).all().total``.
+SUBQUERY_WRAP_PATTERN = re.compile(
+    r"SELECT\s+count\(\*\).*\bFROM\s*\(\s*SELECT\b",
+    re.IGNORECASE | re.DOTALL,
+)
+
+LIMIT_PATTERN = re.compile(r"\bLIMIT\s+(\d+)", re.IGNORECASE)
```
[the remark this answers is not in the document]
Comment (subhashb) [reply src/protean/integrations/pytest/query_assertions.py]: Fixed — this was a real bug: SQLAlchemy emits LIMIT as a bound parameter (`LIMIT ?` on SQLite, `LIMIT %(param)s` on PostgreSQL), so the literal-only regex matched nothing on real ORM queries. `assert_no_overfetch` now resolves the LIMIT value from the captured `parameters` (named and positional), verified to catch over-fetch on both SQLite and PostgreSQL, with tests for every paramstyle.

File: src/protean/integrations/pytest/query_assertions.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    for statement in statements:
+        # Check every LIMIT in the statement, not just the first: an outer
+        # over-fetch can sit behind a smaller inner-subquery limit.
+        for limit in (int(n) for n in LIMIT_PATTERN.findall(statement)):
+            if limit > threshold:
+                offenders.append((statement, limit))
+
+    if offenders:
+        raise AssertionError(
+            f"Over-fetch detected (expected <= {int(threshold)} rows):\n"
```
[the remark this answers is not in the document]
Comment (subhashb) [reply src/protean/integrations/pytest/query_assertions.py]: Fixed — the threshold is now formatted with `:g` (e.g. `1.5`) instead of `int(threshold)`, so the allowed limit is reported accurately.

File: docs/guides/testing/query-shape-tests.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+A `LIMIT 10` passes (`10 <= 10 * 1.5`); a `min(limit * 3, 1000)` over-fetch that
+emits `LIMIT 30` fails. Widen `ratio` when a path legitimately reads a small
+multiple of what it returns.
+
+## Engine resolution
+
+By default the engine is resolved from the active domain: the default
+provider's engine is preferred, falling back to the first SQLAlchemy provider
+found. For tests that manage their own domain or target a non-default provider,
+pass the engine explicitly:
```
[the remark this answers is not in the document]
Comment (subhashb) [reply docs/guides/testing/query-shape-tests.md]: Fixed the docs to match the implementation: resolution uses the active domain's default provider only; for a non-default provider, pass `engine=` explicitly. (The earlier multi-provider fallback was removed because it could resolve the wrong engine.)

File: tests/integrations/pytest/test_query_assertions.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
+"""Tests for the query-shape assertion primitives.
+
+The core logic runs against an in-process SQLite engine passed explicitly, so
+these are plain core tests (no Docker, no adapter marker). Engine resolution
+from the active domain and the no-op behaviour on non-SQLAlchemy backends are
+covered separately.
+"""
+
+import pytest
+from sqlalchemy import create_engine, text
```
[the remark this answers is not in the document]
Comment (subhashb) [reply tests/integrations/pytest/test_query_assertions.py]: Added `pytest.importorskip("sqlalchemy")` at the top of the module so the suite can be collected in environments without SQLAlchemy.

File: tests/integrations/pytest/test_query_assertions.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    eng.dispose()
+
+
+class TestAssertQueryCount:
+    def test_passes_on_exact_count(self, engine):
+        with engine.connect() as conn:
+            with assert_query_count(2, engine=engine):
+                conn.execute(text("SELECT 1"))
+                conn.execute(text("SELECT 2"))
+
```
[the remark this answers is not in the document]
Comment (subhashb) [reply tests/integrations/pytest/test_query_assertions.py:46]: Added `test_pragma_and_transaction_statements_are_not_counted`, which issues a PRAGMA plus a SELECT inside `assert_query_count(1)` to lock in the `_is_query` exclusion.
